tests/mock/clitools.rs

```diff
@@ -221,19 +232,50 @@ pub fn setup_test_state(test_dist_dir: tempfile::TempDir) -> (tempfile::TempDir,
         homedir,
         rustup_update_root: None,
         workdir: RefCell::new(workdir),
+        test_root_dir: test_dir.path().to_path_buf(),
     };
 
-    let build_path = exe_dir.join(format!("rustup-init{EXE_SUFFIX}"));
+    let build_path = built_exe_dir.join(format!("rustup-init{EXE_SUFFIX}"));
 
     let rustup_path = config.exedir.join(format!("rustup{EXE_SUFFIX}"));
-    let setup_path = config.exedir.join(format!("rustup-init{EXE_SUFFIX}"));
+    // Used to create dist servers. Perhaps should only link when needed?
+    let init_path = config.exedir.join(format!("rustup-init{EXE_SUFFIX}"));
     let rustc_path = config.exedir.join(format!("rustc{EXE_SUFFIX}"));
     let cargo_path = config.exedir.join(format!("cargo{EXE_SUFFIX}"));
     let rls_path = config.exedir.join(format!("rls{EXE_SUFFIX}"));
     let rust_lldb_path = config.exedir.join(format!("rust-lldb{EXE_SUFFIX}"));
 
-    copy_binary(build_path, &rustup_path).unwrap();
-    hard_link(&rustup_path, setup_path).unwrap();
+    const ESTIMATED_LINKS_PER_TEST: usize = 6 * 2;
+    // NTFS has a limit of 1023 links per file; test setup creates 6 links, and
+    // then some tests will link the cached installer to rustup/cargo etc,
+    // adding more links
+    const MAX_TESTS_PER_RUSTUP_EXE: usize = 1023 / ESTIMATED_LINKS_PER_TEST;
+    // This returning-result inner structure allows failures without poisoning
+    // cmd_lock.
+    {
+        fn link_or_copy(
+            original: &Path,
+            link: &Path,
+            lock: &mut RwLockWriteGuard<usize>,
+        ) -> io::Result<()> {
+            **lock += 1;
+            if **lock < MAX_TESTS_PER_RUSTUP_EXE {
+                hard_link(original, link)
+            } else {
+                // break the *original* so new tests form a new distinct set of
+                // links. Do this by copying to link, breaking the source,
+                // linking back.
+                **lock = 0;
+                fs::copy(original, link)?;
+                fs::remove_file(original)?;
+                hard_link(link, original)
+            }
+        }
+        let mut _lock = CMD_LOCK.write().unwrap();
```

Comment:
```suggestion
        let mut lock = CMD_LOCK.write().unwrap();
```
We use it below. So probably we should not have a `_` prefix.

Reply:
good point, I'll tweak this.